### ISSUE TITLE
Salted / domain-separated multihash

### ISSUE THREAD

**Issue by nmushegian** [multiformats/multiformats#61]
## background

IPFS already has a concept of multiple CIDs per blob, via multihash. Multibase then gives a second layer of 'user bits' that can be used to intentionally change the CID.

## motivation

There is an overwhelming temptation to intentionally change CIDs in some situations.
Switching hashes and bases is a hack with limited space.
*Letting users invent their own CID domain scheme by salting the content before taking the CID is much worse, because it might interfere with IPFS's native de-duplication*.

## proposal

Allocate one entry in the multihash table to be a 'salted' modifier.
If a hash `h` is assigned prefix `k` (like `k<hash>`) and the salt modifier is assigned prefix `s`, then `sk` is a multihash prefix where `sk<salt><saltedhash>` maps to content `c` such that `saltedhash == h(s, h(c))`.

In other words, instead of adding new hash domains to multiformats by extending the core multihash table, we can add new hash domains to multiformats by adding a standard way of salting the content.

## related work

See use of DOMAIN_SEPARATOR in [EIP712](https://eips.ethereum.org/EIPS/eip-712). Not necessarily the same use case but it shows how giving us a salt will let us do structured domain-separated hashes however we need

**Comment by Stebalien** [comment]
The motivation is pretty slim here. _Why_ do you want to change the CID?

**Comment by nmushegian** [comment]
Good question, I forgot
It was something to do with HMAC and bloom filters and how much you have to store and tyranny of defaults but can't recall the details
I will reopen if I remember
